src/libmeasurement_kit/neubot/negotiate_impl.hpp

```diff
@@ -0,0 +1,213 @@
+        std::cout << "Too many negotiations\n";
+        transport->close([=]() { cb(TooManyNegotiationsError()); });
+        return;
+    };
+
+    request_sendrecv(
+        transport, settings,
+        {
+            {"Content-Type", "application/json"}, {"Authorization", ""},
+        },
+        body,
+        [=](Error error, Var<Response> res) {
+            if (error || (res->status_code != 200)) {
+                std::cout << "Error: " << (int)error;
+                cb(HttpRequestFailedError());
+                return;
+            }
+
+            auto respbody = nlohmann::json::parse(res->body);
+            std::string auth = respbody.at("authorization");
+            auto queue_pos = respbody.at("queue_pos");
+            auto real_address = respbody.at("real_address");
+            int unchoked = respbody.at("unchoked");
+
+            // XXX
+            if (unchoked == 0) {
+                reactor->call_soon([=]() {
+                    loop_negotiate(transport, cb, settings, reactor, logger,
+                                   iteration + 1);
+                });
+
+            } else {
+                dash::run(
+                    settings,
+                    [=](Error err, Var<Entry> measurements) {
+                        if (err) {
+                            logger->warn("Error: %d", (int)error);
+                            cb(err);
+                            return;
+                        }
+
+                        collect(transport,
+                                [=](Error err) {
+                                    if (err) {
+                                        logger->warn("Error: %d", (int)error);
+                                        cb(err);
+                                        return;
+                                    }
+
+                                    cb(NoError());
+                                },
+                                auth, measurements, settings, reactor, logger);
+
+                    },
+                    auth);
+            }
+
+        },
+        reactor, logger);
+}
+
+template <MK_MOCK_NAMESPACE(mlabns, query)>
+void run_impl(Callback<Error> cb, Settings settings, Var<Reactor> reactor,
+              Var<Logger> logger) {
+
+    if (settings["url"] != "") {
+
+        settings["http/url"] = settings["url"];
+
+        if (settings["negotiate"] == "false") {
+            dash::run(settings, [=](Error error, Var<Entry>) {
+                if (error) {
+                    logger->warn("Error: %d", (int)error);
+                    cb(error);
+                    return;
+                }
+            });
+            return;
+        }
+
+        request_connect(settings,
+                        [=](Error error, Var<Transport> transport) {
+
+                            if (error) {
+                                logger->warn("Error: %d", (int)error);
+                                transport->close([=]() { cb(error); });
```

Comment:
I am not sure here we should close the transport. It will most likely be `nullptr`.